src/pe_source/data/shodan/shodan_search.py

```diff
@@ -0,0 +1,420 @@
+import pandas as pd
+import requests
+import shodan
+
+# cisagov Libraries
+from pe_source.data.pe_db.db_query_source import (
+    get_data_source_uid,
+    get_ips,
+    insert_shodan_data,
+)
+
+LOGGER = logging.getLogger(__name__)
+
+
+def run_shodan_thread(api, org_chunk, thread_name):
+    """Run a Shodan thread."""
+    failed = []
+    for org in org_chunk:
+        org_name = org["cyhy_db_name"]
+        org_uid = org["org_uid"]
+        LOGGER.info("{} Running IPs for {}".format(thread_name, org_name))
+        start, end = get_dates()
+        try:
+            ips = get_ips(org_uid)
+        except Exception as e:
+            LOGGER.error("{} Failed fetching IPs for {}.".format(thread_name, org_name))
+            LOGGER.error("{} {} - {}".format(thread_name, e, org_name))
+            failed.append("{} fetching IPs".format(org_name))
+            continue
+
+        if len(ips) == 0:
+            LOGGER.error("{} No IPs for {}.".format(thread_name, org_name))
+            failed.append("{} has 0 IPs".format(org_name))
+            continue
+
+        failed = search_shodan(
+            thread_name, ips, api, start, end, org_uid, org_name, failed
+        )
+
+    if len(failed) > 0:
+        LOGGER.critical("{} Failures: {}".format(thread_name, failed))
+
+
+def get_dates():
+    """Get dates for the query."""
+    now = datetime.datetime.now()
+    days_back = datetime.timedelta(days=30)
+    days_forward = datetime.timedelta(days=1)
+    start = now - days_back
+    end = now + days_forward
+    start_time = time_to_utc(start)
+    end_time = time_to_utc(end)
+    return start_time, end_time
+
+
+def time_to_utc(in_time):
+    """Convert time to UTC."""
+    # If time does not have timezone info, assume it is local
+    if in_time.tzinfo is None:
+        local_tz = datetime.datetime.now().astimezone().tzinfo
+        in_time = in_time.replace(tzinfo=local_tz)
+    utc_time = in_time.astimezone(datetime.timezone.utc)
+    return utc_time
+
+
+def search_circl(cve):
+    """Fetch CVE info from Circl."""
+    re = requests.get(f"https://cve.circl.lu/api/cve/{cve}")
+    return re
+
+
+def search_shodan(thread_name, ips, api, start, end, org_uid, org_name, failed):
+    """Search IPs in the Shodan API."""
+    # Initialize lists to store Shodan results
+    data = []
+    risk_data = []
+    vuln_data = []
+
+    # Build dictionaries for naming conventions and definitions
+    risky_ports, name_dict, risk_dict, av_dict, ac_dict, ci_dict = get_shodan_dicts()
+
+    # Break up IPs into chunks of 100
+    tot_ips = len(ips)
+    ip_chunks = [ips[i : i + 10] for i in range(0, tot_ips, 10)]
+    tot = len(ip_chunks)
+    LOGGER.info(
+        "{} Split {} IPs into {} chunks - {}".format(
+            thread_name, tot_ips, tot, org_name
+        )
+    )
+
+    # Loop through chunks and query Shodan
+    for i, ip_chunk in enumerate(ip_chunks):
+        count = i + 1
+        try_count = 1
+        while try_count < 7:
```

Comment:
Magic numbers like `7` should be avoided; use variables so you can easily change the value later if necessary.

Reply:
Fixed here: [10f0e6d](https://github.com/cisagov/pe-source/pull/1/commits/10f0e6d6e5c97425615bd3c057aa9040ecbf8928)